src/backend/opengl/PersistentPipelineStateGL.cpp

```diff
@@ -0,0 +1,169 @@
+// Copyright 2017 The NXT Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "PersistentPipelineStateGL.h"
+
+#include "OpenGLBackend.h"
+
+namespace backend {
+namespace opengl {
+
+    PersistentPipelineState::PersistentPipelineState() {
+        dirtyFields.set(); // initialize all fields as dirty
+    }
+
+    // when a field on PersistentPipelineState::State changes, mark its starting location as dirty
+    #define SET_FIELD(field, destination, value) { \
+        if (state.destination != value) { \
+            dirtyFields.set(field); \
+            state.destination = value; \
+        } \
+    }
+
+    void PersistentPipelineState::UpdateDepthStencilInfo(const DepthStencilStateBase* const depthStencilState) {
```

Comment:
nit: we generally don't need constant pointers, only pointers to constant data (and the function is free to point to a different object if it wants to). (aka you can remove the second const)